File: EPI/12.12/swift/12.12_test.swift
import Darwin

let tests: [(numbers: [Int], duplicate: Int, missing: Int)] = [
    (
        numbers: [0, 0],
        duplicate: 0,
        missing: 1
    ),
    (
        numbers: [4, 5, 2, 9, 7, 0, 1, 5, 6, 8],
        duplicate: 5,
        missing: 3
    ),
    (
        numbers: [4, 5, 2, 9, 7, 0, 1, 3, 2, 8],
        duplicate: 2,
        missing: 6
    ),
]

for test in tests {
    let actual = findDuplicateAndMissing(test.numbers)
    let expected = (duplicate: test.duplicate, missing: test.missing)
    guard actual == expected else {
        print("For test numbers \(test.numbers), " +
            "expected (duplicate, missing) to be \(expected), but was \(actual)")
        exit(1)
    }
}

File: EPI/Searching/Find the duplicate and missing elements/typescript/12.12_test.ts
import findDuplicateAndMissing from './12.12';

declare function require(name: string): any;
const assert = require('assert');

const tests: {values: number[], duplicate: number, missing: number}[] = [
    {
        values: [0, 0],
        duplicate: 0,
        missing: 1
    },
    {
        values: [4, 5, 2, 9, 7, 0, 1, 5, 6, 8],
        duplicate: 5,
        missing: 3
    },
    {
        values: [4, 5, 2, 9, 7, 0, 1, 3, 2, 8],
        duplicate: 2,
        missing: 6
    },
];

for (const test of tests) {
    const {duplicate, missing} = findDuplicateAndMissing(test.values);
    assert.deepStrictEqual([duplicate, missing], [test.duplicate, test.missing],
        `For values ${test.values}, ` +
        `expected duplicate to be ${test.duplicate} and missing to be ${test.missing}, ` +
        `but duplicate was ${duplicate} and missing was ${missing}`);
}
